IBPSA/Fluid/FixedResistances/CheckValve.mo

```diff
@@ -27,48 +27,75 @@ model CheckValve "Check valve that avoids flow reversal"
-
-  Real k(min=Modelica.Constants.small)
-    "Flow coefficient of valve and pipe in series in allowed/forward direction,
-    k=m_flow/sqrt(dp), with unit=(kg.m)^(1/2).";
-
 protected
-  Real a
-    "Scaled pressure variable";
-  Real cv
-    "Twice differentiable Heaviside check valve characteristic";
-  Real kCv
-    "Smoothed restriction characteristic";
-
+  parameter Real k_min=if dpFixed_nominal > Modelica.Constants.eps then
+    sqrt(1 / (1 / kFixed ^ 2 + 1 /(l * Kv_SI) ^ 2)) else l * Kv_SI
+    "Minimum flow coefficient (valve closed)";
+  parameter Real k_max=if dpFixed_nominal > Modelica.Constants.eps then
+    sqrt(1 / (1 / kFixed ^ 2 + 1 / Kv_SI ^ 2)) else Kv_SI
+    "Maximum flow coefficient (valve fully open)";
+  Modelica.Units.SI.MassFlowRate m_flow_min
+    "Flow rate through closed valve";
+  Modelica.Units.SI.MassFlowRate m_flow_max
+    "Flow rate through fully open valve";
+  Modelica.Units.SI.MassFlowRate m_flow_smooth
+    "Smooth interpolation result between two flow regimes";
 initial equation
-  assert(dpFixed_nominal > -Modelica.Constants.eps,
-    "In " + getInstanceName() + ": We require dpFixed_nominal >= 0.
+  assert(dpFixed_nominal > - Modelica.Constants.eps, "In " + getInstanceName() +
+    ": We require dpFixed_nominal >= 0.
     Received dpFixed_nominal = " + String(dpFixed_nominal) + " Pa.");
-  assert(l > -Modelica.Constants.eps,
-    "In " + getInstanceName() + ": We require l >= 0. Received l = " + String(l));
+  assert(l > - Modelica.Constants.eps, "In " + getInstanceName() +
+    ": We require l >= 0. Received l = " + String(l));
 equation
-  a = dp/dpValve_closing;
-  cv = smooth(2, max(0, min(1, a^3*(10+a*(-15+6*a)))));
-  kCv = Kv_SI*(cv*(1-l) + l);
-
-  if (dpFixed_nominal > Modelica.Constants.eps) then
-    k = sqrt(1/(1/kFixed^2 + 1/kCv^2));
-  else
-    k = kCv;
-  end if;
-
+  // min function ensures that m_flow_min does not increase further for dp > 0
+  m_flow_min = IBPSA.Fluid.BaseClasses.FlowModels.basicFlowFunction_dp(
+    dp=min(dp, 0),
+    k=k_min,
+    m_flow_turbulent=m_flow_turbulent);
+  // max function ensures that m_flow_max does not decrease further for dp < dpValve_closing
+  m_flow_max = IBPSA.Fluid.BaseClasses.FlowModels.basicFlowFunction_dp(
+    dp=max(dp, dpValve_closing),
+    k=k_max,
+    m_flow_turbulent=m_flow_turbulent);
+  // We compute d(m_flow)/dp by calling the function d(m_flow)/dt with dp/dt=1,
+  // and d2(m_flow)/dp2 by calling the function d2(m_flow)/dt2 with dp/dt=1 and d2p/dt2=0
+  m_flow_smooth=noEvent(smooth(2,
+    if dp <= 0 then m_flow_min
+    elseif dp >= dpValve_closing then m_flow_max
+    else IBPSA.Utilities.Math.Functions.quinticHermite(
+      x=dp,
+      x1=0,
+      x2=dpValve_closing,
+      y1=m_flow_min,
+      y2=m_flow_max,
+      y1d=IBPSA.Fluid.BaseClasses.FlowModels.basicFlowFunction_dp_der(
+        dp=0,
+        k=k_min,
+        m_flow_turbulent=m_flow_turbulent,
+        dp_der=1),
```

Comment:
Can you explain why `dp_der=1` and `dp_der2=0`. I don't understand what you implemented here.
(A side note is that these are parameter expressions so for efficiency that could be computed as a parameter.)

Reply:
See above.
For the second derivative, considering
$`\frac{\partial^2 m\_flow}{\partial t^2} = \frac{\partial^2 m\_flow}{\partial dp^2} * \left(\frac{\partial dp}{\partial t}\right)^2 + \frac{\partial m\_flow}{\partial dp} * \frac{\partial^2 dp}{\partial t^2}`$
I compute $`\frac{\partial^2 m\_flow}{\partial dp^2}`$ by calling `basicFlowFunction_dp_der2(..., dp_der=1, dp_der2=0)`.